tests/Validation/NoUndefinedVariablesRuleTest.hack

```diff
@@ -0,0 +1,343 @@
+                    ),
+                ],
+            ),
+            'variable in fragment used by other operation' => tuple(
+                '
+                query Foo($b: String) {
+                    ...FragA
+                }
+                query Bar($a: String) {
+                    ...FragB
+                }
+                fragment FragA on Type {
+                    field(a: $a)
+                }
+                fragment FragB on Type {
+                    field(b: $b)
+                }
+                ',
+                vec[
+                    shape(
+                        'message' => 'Variable "$a" is not defined by operation "Foo".',
+                        'location' => shape('line' => 9, 'column' => 28),
+                    ),
+                    shape(
+                        'message' => 'Variable "$b" is not defined by operation "Bar".',
+                        'location' => shape('line' => 12, 'column' => 28),
+                    ),
+                ],
+            ),
+            'multiple undefined variables produce multiple errors' => tuple(
+                '
+                query Foo($b: String) {
+                    ...FragAB
+                }
+                query Bar($a: String) {
+                    ...FragAB
+                }
+                fragment FragAB on Type { 
+                    field1(a: $a, b: $b)
+                    ...FragC
+                    field3(a: $a, b: $b)
+                }
+                fragment FragC on Type {
+                    field2(c: $c)
+                }
+                ',
+                vec[
+                    shape(
+                        'message' => 'Variable "$a" is not defined by operation "Foo".',
+                        'location' => shape('line' => 9, 'column' => 29),
+                    ),
+                    shape(
+                        'message' => 'Variable "$a" is not defined by operation "Foo".',
+                        'location' => shape('line' => 11, 'column' => 29),
+                    ),
+                    shape(
+                        'message' => 'Variable "$c" is not defined by operation "Foo".',
+                        'location' => shape('line' => 14, 'column' => 29),
+                    ),
+                    shape(
+                        'message' => 'Variable "$b" is not defined by operation "Bar".',
+                        'location' => shape('line' => 9, 'column' => 36),
+                    ),
+                    shape(
+                        'message' => 'Variable "$b" is not defined by operation "Bar".',
+                        'location' => shape('line' => 11, 'column' => 36),
+                    ),
+                    shape(
+                        'message' => 'Variable "$c" is not defined by operation "Bar".',
+                        'location' => shape('line' => 14, 'column' => 29),
+                    ),
+                ],
+            ),
```

Comment:
Maybe add a test with a fragment loop.

```GraphQL
fragment Foo {
  ...Foo
}
```

There is another validation rule to detect these and output errors, but we should make sure it doesn't cause us to go into an infinite loop in this rule or elsewhere.

Reply:
Added — it passes.